crates/core/tedge_api/src/entity_store.rs

```diff
@@ -191,15 +186,109 @@ impl EntityStore {
             twin_data: Map::new(),
         };
 
-        Some(EntityStore {
+        let message_log = MqttMessageLog::new(log_dir)?;
+
+        let mut entity_store = EntityStore {
+            mqtt_schema: mqtt_schema.clone(),
             main_device: main_device.topic_id.clone(),
             entities: HashMap::from([(main_device.topic_id.clone(), metadata)]),
             entity_id_index: HashMap::from([(entity_id, main_device.topic_id)]),
             external_id_mapper: Box::new(external_id_mapper_fn),
             external_id_validator_fn: Box::new(external_id_validator_fn),
             default_service_type,
             pending_entity_store: PendingEntityStore::new(mqtt_schema, telemetry_cache_size),
-        })
+            message_log,
+        };
+
+        entity_store.load_from_message_log();
+
+        Ok(entity_store)
+    }
+
+    pub fn load_from_message_log(&mut self) {
+        info!("Loading the entity store from the log");
+        match self.message_log.reader() {
```

Comment:
I would be better to transform this method into a function taking mutable references to an entity store and a log of messages; populating the store from these messages.

A bit more involved, but I would also try to persist the entity store from outside, i.e with a function taking the store and the message log. This can be done by using an internal vector of non-persisted-yet messages.

Doing so:
- The entity store will stay a pure data structure with no side effects.
- The load and persist methods can be made async.
- The logic would be simpler avoiding this usage of boolean flags to control in depth when something needs to be persisted or not.
- The coupling to the mqtt_schema could also be removed that way.

Reply:
As discussed offline, this would be deferred for later.